[FLINK-8577][table] Implement proctime DataStream to Table upsert conversion

hequn8128 opened the pull request:
## What is the purpose of the change

Currently, only append stream can be ingested as a stream table. This pull request implement proctime DataStream to Table upsert conversion.

Api looks like:

```
DataStream[(Boolean, (String, Long, Int))] input = ???
// upsert with keyedTable 
table = tEnv.fromUpsertStream(input, 'a, 'b, 'c.key)
// upsert without key -> single row tableTable 
table = tEnv.fromUpsertFromStream(input, 'a, 'b, 'c) 
```

A simple design doc can be fond [here](https://docs.google.com/document/d/1yMEi9xkm3pNYNmmZykbLzqtKjFVezaWe0y7Xqd0c1zE/edit?usp=sharing).


## Brief change log

  - Add `fromUpsertStream()` and `fromAppendStream()` in  java and scala `StreamTableEnvironment` and deprecate `fromDataStream`.
  - Add key support in the source definition including parse `key` keyword. 
  - Add `DataStreamLastRowRule` and `DataStreamLastRowAfterCalcRule`. Both rules generate `DataStreamLastRow` to handle upsert stream. The differences between the two rule is `DataStreamLastRowAfterCalcRule` will take calc into consideration and generate LastRow DataStreamRel node after calc. This can decrease state size in LastRow.
  - Add `LastRowProcessFunction` to handle upsert messages and generate retractions if there is an update.


## Verifying this change

This change added tests and can be verified as follows:

  - Add java api test in `JavaSqlITCase`
  - Add IT test cases in `FromUpsertStreamITCase`
  - Add sql plan tests in `FromUpsertStreamTest`
  - Add key extract in `UpdatingPlanCheckerTest`
  - Add validation test in `StreamTableEnvironmentValidationTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented, documents will be added in the later pr)

hequn8128 commented:
Discussed with Fabian in [FLINK-8578](https://issues.apache.org/jira/browse/FLINK-8578), we have to materialize the proc-time field and convert row-time attributes into regular TIMESTAMP attributes. I will update the PR ASAP.

hequn8128 commented:
Update pr with the following changes: 1. materialize time indicators 2. refactor rules

hequn8128 commented:
@pnowojski  Hi, would be great if you can take a look when you are free. Thank you.

pnowojski commented:
Sorry @hequn8128 that you have to wait that long and I remember about this PR, but I was busy with 1.7 release. I hopefully will be able to get back to this one in the next couple of days.

hequn8128 commented:
Thanks, @pnowojski ! Is there anything I can help for the release?

sunjincheng121 commented:
@hequn8128 Thanks for the PR. 
I am glad to review the changes after you rebase the code!
Best,
Jincheng

hequn8128 commented:
@sunjincheng121  thanks for your attention. I have rebased to the master.

dianfu commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/conversion/JavaTupleToCRowProcessRunner.scala:
```diff
@@ -0,0 +1,75 @@
+/**
+  * ProcessRunner with [[CRow]] output.
+  */
+class JavaTupleToCRowProcessRunner(
```
Seems that we can reuse the ExternalTypeToCRowProcessRunner/CRowOutputProcessRunner. Just need to do some checks whether the input is java Tuple, scala tuple or other types.

hequn8128 replied:
Doing checks will influence the performance. I think we can generate code according to the type to solve the problem. However, we can improve this in another pr. What do you think?

dianfu replied:
Make sense.

dianfu commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/conversion/JavaTupleToCRowProcessRunner.scala:
```diff
@@ -0,0 +1,75 @@
+    LOG.debug(s"Compiling ProcessFunction: $name \n\n Code:\n$code")
+    val clazz = compile(getRuntimeContext.getUserCodeClassLoader, name, code)
+    LOG.debug("Instantiating ProcessFunction.")
+    function = clazz.newInstance()
```
Add the open/close logic such as FunctionUtils.setFunctionRuntimeContext/FunctionUtils.openFunction/FunctionUtils.closeFunction

dianfu commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/conversion/ScalaTupleToCRowProcessRunner.scala:
```diff
@@ -0,0 +1,72 @@
+    LOG.debug(s"Compiling ProcessFunction: $name \n\n Code:\n$code")
+    val clazz = compile(getRuntimeContext.getUserCodeClassLoader, name, code)
+    LOG.debug("Instantiating ProcessFunction.")
+    function = clazz.newInstance()
```
Add the open/close logic such as FunctionUtils.setFunctionRuntimeContext/FunctionUtils.openFunction/FunctionUtils.closeFunction

dianfu commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/FromUpsertStreamITCase.scala:
```diff
@@ -0,0 +1,207 @@
+import org.junit.Assert._
+import org.junit._
+
+class FromUpsertStreamITCase extends StreamingWithStateTestBase {
```
Rename to UpsertStreamITCase?

hequn8128 replied:
UpsertStream sounds like Upsert a stream into a sink. What do you think?

dianfu commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/FromUpsertStreamTest.scala:
```diff
@@ -0,0 +1,140 @@
+import org.apache.flink.api.java.tuple.{Tuple2 => JTuple2}
+import java.lang.{Boolean => JBool}
+
+class FromUpsertStreamTest extends TableTestBase {
```
Rename to UpsertStreamTest?

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/RelTimeIndicatorConverter.scala:
```diff
@@ -165,7 +165,30 @@ class RelTimeIndicatorConverter(rexBuilder: RexBuilder) extends RelShuttle {
+        scan
+      }
+
+      LogicalLastRow.create(
```
Year, It would be nice if we can. A rule can only transform from a node to a new node with a same rowtype. However, the rowtype of LastRow is different from the scan, i.e, the tuple type has been flattened. I suggest to

pnowojski replied:
But you could probably add this node when changing convention. We assume that `TabelScan` rel node represents all kinds of table scans and it always can produce retractions, while for some stage (`FlinkLogical`?) we freeze it to either retractions/upserts.

Regardless of that, `RelTimeIndicatorConverter.scala` is not the right place to create `LogicalLastRow`. You could also add a single pass rule, that is applied only once, that inserts upsert -> retraction conversion.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamLastRow.scala:
```diff
@@ -0,0 +1,122 @@
+    val needRetraction = DataStreamRetractionRules.isAccRetract(this)
+    val result: DataStream[CRow] = if (needRetraction) {
+      val processFunction = new LastRowProcessFunction(
+        new RowTypeInfo(schema.fieldTypeInfos.toArray, schema.fieldNames.toArray),
```
`LastRowProcessFunction` requires a type of RowTypeInfo while `schema.typeInfo` returns a type of TypeInformation[Row]

dianfu replied:
schema.typeInfo can be cast to RowTypeInfo as it's always RowTypeInfo.

hequn8128 replied:
You are right.

pnowojski commented:
I've started reviewing the code and as we talked offline, could you split this PR into 4 commits:
- Add `fromAppendStream` and Deprecate `fromDataStream`
- Introduce `LastRow` Rules
- Introduce `LastRow` runtime functions
- Optimize upsert sources followed by `Calc`

? Having more, smaller, independent commits speeds up reviewing and help in the future whenever someone is digging through git's commit history to understand why/how was something implemented. Especially splitting of refactoring/renaming commits from feature adding/changing or bug fixing commits is important.

hequn8128 commented:
@pnowojski  thanks for your review. Make sense to me. I will split it into 4 commits according to your suggestions. Meanwhile, I will address the comments from @dianfu . Thank you all.

hequn8128 commented:
@pnowojski @sunjincheng121 @dianfu Thanks for your review and suggestions. I have split and updated the pr.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/DataStreamConversions.scala:
```diff
@@ -59,5 +59,38 @@ class DataStreamConversions[T](dataStream: DataStream[T], inputType: TypeInforma
+    * @param fields The field names of the new [[Table]] (optional).
+    * @return The resulting [[Table]].
+    */
+  def toKeyedTable(tableEnv: StreamTableEnvironment, fields: Expression*): Table = {
```
This is some naming inconsistency: `toKeyedTable` vs `fromUpsertStream`. Rename `keyed` to `upsert`?

hequn8128 replied:
I think we can't use upsert to describe a Table. How about rename it to `toTableFromUpsert`?

pnowojski replied:
Maybe `upsertStreamToTable`?

hequn8128 replied:
I prefer the name starts with toTable. So the methods would be: `toTable`, `toTableFromUpsertStream`, `toTableFromRetractStream`. We can find these methods easier, similar to `fromUpsertStream `, `fromAppendStream `, `fromDataSet `. What do you think?

hequn8128 replied:
I think the difference between `toTableFromRetractStream` and `toTable` is `toTable` means `toTableFromAppendStream`. The method names are consistence.
Methods in `StreamTableEnvironment`:
- fromAppendStream
- fromUpsertStream
- fromRetractStream

Methods in `DataStreamConversions`:
- toTableFromAppendStream
- toTableFromUpsertStream
- toTableFromRetractStream

Renaming `fromDataStream` to `fromAppendStream` is consistent with renaming `toDataStream` to `toAppendStream` by [FLINK-6543](https://issues.apache.org/jira/browse/FLINK-6543)

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/RelTimeIndicatorConverter.scala:
```diff
@@ -165,7 +165,30 @@ class RelTimeIndicatorConverter(rexBuilder: RexBuilder) extends RelShuttle {
+        scan
+      }
+
+      LogicalLastRow.create(
```
Do we ALWAYS convert upserts to retractions? Even for pipelines `upsert source -> filter -> upsert sink` ?

hequn8128 replied:
No. As we discussed before, `LastRow` will become a no-op for the case of `upsert source -> filter -> upsert sink`.

hequn8128 replied:
You are right. 
Currently, there is no such no-op node in the execution plan. If you still concern with the no-op node in the optimization plan, I can add a rule in the RetractionRules to remove it. In this way, we can name `LastRow` to `UpsertToRetractionConverter`. What do you think?

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/ExpressionParser.scala:
```diff
@@ -535,6 +536,13 @@ object ExpressionParser extends JavaTokenParsers with PackratParsers {
       case f ~ _ ~ _ => RowtimeAttribute(f)
     }
 
+  // key
```
nit: this comment is not helpful it just duplicates the field name

hequn8128 replied:
Ok, I will remove it.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/ExpressionParser.scala:
```diff
@@ -535,6 +536,13 @@ object ExpressionParser extends JavaTokenParsers with PackratParsers {
+  // key
+
+  lazy val key: PackratParser[Expression] =
+    (aliasMapping | "(" ~> aliasMapping <~ ")" | fieldReference) ~ "." ~ KEY ^^ {
```
can we deduplicate `(aliasMapping | "(" ~> aliasMapping <~ ")" | fieldReference)`? I would expect so but I'm not familiar with this scala magic `ExpressionParser`.

hequn8128 replied:
Do you mean deduplicate the code? I think we can add another `Expression`, e.g., `lazy val aliasOrFieldReference = aliasMapping | "(" ~> aliasMapping <~ ")" | fieldReference`, and reuse it.

pnowojski replied:
Yes, I meant that this `(aliasMapping | "(" ~> aliasMapping <~ ")" | fieldReference)` is duplicated couple of times here. If it is possible to deduplicate it to:
```
lazy val aliasOrFieldReference = aliasMapping | "(" ~> aliasMapping <~ ")" | fieldReference
```
I think it would be better to do so.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/AppendStreamScan.scala:
```diff
@@ -36,20 +36,20 @@ import org.apache.flink.table.typeutils.TimeIndicatorTypeInfo
   * the plan translation.
   */
-class DataStreamScan(
+class AppendStreamScan(
```
Shouldn't this belong to the previous commit?

hequn8128 replied:
Yes, sorry about this.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/logical/rel/LogicalLastRow.scala:
```diff
@@ -0,0 +1,54 @@
+import org.apache.calcite.rel.{RelNode, SingleRel}
+
+/**
+  * Represent an upsert source.
```
? I think this comment is incorrect. This is not the source on it's own, but only a conversion class.

hequn8128 replied:
I treat `LastRow` and `Source` as an entirety. However, I think you are right. Treat `LastRow` as a conversion node may be better.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/logical/rel/LogicalLastRow.scala:
```diff
@@ -0,0 +1,54 @@
+  *
+  * @param keyNames   The upsert key names.
+  */
+class LogicalLastRow(
```
I still think this is a bad and misleading name for this operator. `LastRow` doesn't explain neither the purpose, nor the actual thing that it's doing. If someone was asked without prior knowledge what `LastRow` RelNode/operator does, it would be impossible to guess the correct answer. It's even more confusing when you are binding `LastRow` with "upsert" name in the java doc:

> Represent an upsert source.

> The upsert key names.

Please either change the name to something like `UpsertToRetractionsConverter` that explicitly states the purpose of this class (than you can keep `upsert` references in the java docs) or if you would like to keep naming by the thing that it does it would have to meet the following requirements:
- `LastRow` is independent of the upsert -> retraction conversion logic
- there is a reasonable chance that it will be used somewhere else outside of the upsert -> retraction conversion context 

It's a bit like we would rename `LogicalJoin` to `HashTable`. Kind of true but misleading. It could make sense if:
- `HashTable` is extracted to separate class
- `LogicalJoin` relnode/operator/concept still exists, but is just using `HashTable`
- `HashTable` is reused somewhere else

hequn8128 replied:
Year, I also think `LastRow` is not perfect. However, I can't find a better one. We can't name it `UpsertToRetractionsConverter ` since we don't always convert upsert to retractions in this node. `LastRow` does explain the purpose and the thing that it is doing, i.e. get the last row from the upsert stream by the defined key.

pnowojski replied:
>  We can't name it UpsertToRetractionsConverter since we don't always convert upsert to retractions in this node. 

can you elaborate on that?

hequn8128 replied:
The `LastRow` node may be a no-op node in the case such as `upsert source -> calc -> upsert sink`. And `LastRow` will convert upsert stream to retract stream if a downstream node needs it to, such as `upsert source -> calc -> retract sink`. Whether convert to retract stream will be decided by RetractionRules.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/logical/rel/LogicalLastRow.scala:
```diff
@@ -0,0 +1,54 @@
+/**
+  * Represent an upsert source.
+  *
+  * @param keyNames   The upsert key names.
```
if `keyNames` are "upsert key names" then just name it `upsertKeyNames` and drop the comment.

hequn8128 commented on flink-examples/flink-examples-table/src/main/scala/org/apache/flink/table/examples/scala/StreamSQLExample.scala:
```diff
@@ -54,9 +54,9 @@ object StreamSQLExample {
       Order(4L, "beer", 1)))
 
     // convert DataStream to Table
-    var tableA = tEnv.fromDataStream(orderA, 'user, 'product, 'amount)
```
I plan to address the problems of document in [FLINK-10957](https://issues.apache.org/jira/browse/FLINK-10957). What do you think?

sunjincheng121 replied:
+1 sync the changes to documentation and also agree the document change can be in an other PR.

pnowojski replied:
I would prefer to document this rename in this PR, but if you want it all do together with upsert sources documentation then OK, but let's not forget about it.

hequn8128 replied:
ok, I will not forget about it. Thank you.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/StreamTableEnvironment.scala:
```diff
@@ -555,18 +555,105 @@ abstract class StreamTableEnvironment(
+    * @param dataStream The [[DataStream]] to register as table in the catalog.
+    * @tparam T the type of the [[DataStream]].
+    */
+  protected def registerUpsertStreamInternal[T](name: String, dataStream: DataStream[T]): Unit = {
```
The primary key would be necessary. It would be a single row table if the primary key has not been defined.

pnowojski replied:
Ok, I get it - upsert streams without a primary key are single row tables. But what do you mean by:
> The primary key would be necessary.

?

hequn8128 replied:
Oh, sorry about my typo. It should be "The primary key would not be necessary."

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/StreamTableEnvironment.scala:
```diff
@@ -555,18 +555,105 @@ abstract class StreamTableEnvironment(
+    // Can not apply key on append stream
+    if (extractUniqueKeys(fields).nonEmpty) {
+      throw new TableException(
+        s"Can not apply key on append stream, use fromUpsertStream instead.")
```
Agree, specifing unique key on an AppendTable makes sense. For example, we only insert data into a table without an update. 
Since supporting keys on AppendStream is another new feature. I think we can add this check now and remove it later if we support keys on AppendStream. The error message should be corrected as `Apply key on append stream has not been supported yet! You can use fromUpsertStream instead.`
What do you think?

pnowojski replied:
I meant that this check doesn't prevent from any bugs or from any other exceptions. It's just to block user from specifying a no-op key definition, right? That's why I'm not sure if we should block it.

If we decide to keep this check/exception, I would refrain from naming `fromUpsertStream` in the exception. Maybe rephrasing it to:
> Defining key on append stream do not have any effects

?

hequn8128 replied:
I haven't thought much about how to support key definition on append stream. I think it may not be a no-op key definition. Maybe we have to do the following things:
- Add similar logic in `UniqueKeyExtractor` to get keys.
- Probably have to check whether data meet the key constraint. i.e, throw an exception when key data is not distinct. 
- etc.
Given the above things. I think we would better to not support it now? 
I will rephrase the error message according to your suggestions. Thanks a lot.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala:
```diff
@@ -1089,6 +1089,19 @@ abstract class TableEnvironment(val config: TableConfig) {
               referenceByName(origName, t).map((_, name))
             }
+          case (Key(UnresolvedFieldReference(name: String)), idx) =>
+            if (isRefByPos) {
```
Same as bellow.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala:
```diff
@@ -1101,6 +1114,10 @@ abstract class TableEnvironment(val config: TableConfig) {
             referenceByName(name, p).map((_, name))
           case Alias(UnresolvedFieldReference(origName), name: String, _) =>
             referenceByName(origName, p).map((_, name))
+          case (Key(UnresolvedFieldReference(name: String))) =>
```
UnresolvedKeyFieldReference seems a good way. I will try to figure it out.

hequn8128 replied:
UnresolvedKeyFieldReference works fine. Thanks for your suggestions.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/DataStreamConversions.scala:
```diff
@@ -59,5 +59,38 @@ class DataStreamConversions[T](dataStream: DataStream[T], inputType: TypeInforma
+  /**
+    * Converts the [[DataStream]] with upsert messages into a [[Table]] with keys.
+    *
+    * The message will be encoded as [[Tuple2]]. The first field is a [[Boolean]] flag, the second
```
The messages in the table will be encoded as `Tuple2`? Or the incoming messages from the source `DataStream` are expected to be encoded as `Tuple2`?

hequn8128 replied:
The incoming messages from the source `DataStream` are expected to be encoded as `Tuple2`. I will change the description according to your suggestion. Thank you.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/logical/FlinkLogicalLastRow.scala:
```diff
@@ -0,0 +1,74 @@
+    val rowCnt = mq.getRowCount(child)
+    // take rowCnt and fieldCnt into account, so that cost will be smaller when generate LastRow
+    // after Calc.
+    planner.getCostFactory.makeCost(rowCnt, rowCnt * child.getRowType.getFieldCount, 0)
```
as mentioned above (in the no-op/renaming discussion), this cost computation is not/will not be correct in case of no-op

hequn8128 replied:
Yes, it would be better if we know it is a no-op node in the logical optimize phase. I think the computation is ok since it is always right to push calc down for the plan. The plan is right. However, it is true that it will introduce some side effect, i.e., with greater search space. 

It is not easy to know whether it is a no-op node in the logical phase since the retraction rules have not yet been performed. Do you have any suggestions?

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamScanRule.scala:
```diff
@@ -36,25 +36,32 @@ class DataStreamScanRule
+    val appendTable = scan.getTable.unwrap(classOf[AppendStreamTable[Any]])
+    val upsertTable = scan.getTable.unwrap(classOf[UpsertStreamTable[Any]])
+
+    appendTable != null || upsertTable != null
```
can we avoid those nulls (in multiple places)? For example by using `isInstanceOf` or match? Handling null is dangerous and error prone.

hequn8128 replied:
I think you are right.

hequn8128 commented:
@pnowojski Hi, thanks a lot for your review and great suggestions. I have updated the pr according to your suggestions. Changes mainly include:
- Add `UnresolvedKeyFieldReference`
- Rename `LastRow` to `UpsertToRetraction`
- Generate `UpsertToRetraction` from `RelTimeIndicatorConverter` to `EnumerableToLogicalTableScan`
- Add `RemoveDataStreamUpsertToRetractionRule` to `UpsertToRetraction` relnode if it is a no-op node.
Would be great if you can take a look when you are free. Thanks a lot.

Best, Hequn

hequn8128 commented:
@pnowojski  Thanks a lot for your review. I have updated the pr according to your suggestions. I split the updates into 4 commits which I think may be helpful for the review. 
I haven't rebased to the master. I prefer to do it once for all when everything is ok. Is it ok with you?

Best, Hequn

hequn8128 commented:
@pnowojski  Hi, should we move on :-)

hequn8128 commented:
@pnowojski Thanks a lot for the view. I have updated the PR according to your suggestions. Mainly include the following changes:
- Add `UpsertToRetractionConverter`
- Resplit the commits
- Rebase to the master
Would be great if you can take another look.

Best, Hequn

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/RelTimeIndicatorConverter.scala:
```diff
@@ -215,6 +217,19 @@ class RelTimeIndicatorConverter(rexBuilder: RexBuilder) extends RelShuttle {
     materializerUtils.projectAndMaterializeFields(rewrittenTemporalJoin, indicesToMaterialize)
   }
 
+  def visit(upsertToRetraction: LogicalUpsertToRetraction): RelNode = {
```
Good catch!
> First you didn't recursively call the RelTimeIndicatorConverter on the upsertToRetraction input
The `LogicalUpsertToRetraction` is right after the source. There is no case we need to materialize time indicators in source. Currently, visit() return scan directly in `RelTimeIndicatorConverter`. But I think you are right. It is better to visit input for `LogicalUpsertToRetraction`. However, it seems impossible for us to add a test for this.

> Secondly, we should solve this in some more generic way.
Agree. I suggest to do it in another pr. It is self contained. This pr is a bit big. I would rather not to add new feature in it. What do you think?

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/UpsertStreamScan.scala:
```diff
@@ -0,0 +1,85 @@
+  * Flink RelNode which matches along with DataStreamSource. Different from [[AppendStreamScan]],
+  * [[UpsertStreamScan]] is used to handle upsert streams from source.
+  */
+class UpsertStreamScan(
```
Seems a common base class is better. What do you think?

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/UpdatingPlanChecker.scala:
```diff
@@ -183,6 +183,17 @@ object UpdatingPlanChecker {
+            .map(_._1)
+          Some(uniqueKeyNames.map(e => (e, e)))
+
+        case scan: UpsertStreamScan =>
```
The `visite` in `RelVisitor` returns void. We need to return an `Option[Seq[(String, String)]]` here

hequn8128 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/StreamTableEnvironmentTest.scala:
```diff
@@ -170,6 +170,19 @@ class StreamTableEnvironmentTest extends TableTestBase {
   }
 
+  @Test
+  def testAddTableFromUpsert(): Unit = {
```
Used to check `registerTable()` with `field.key`

hequn8128 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/StreamTableEnvironmentTest.scala:
```diff
@@ -185,4 +198,20 @@ class StreamTableEnvironmentTest extends TableTestBase {
+  private def prepareKeyedSchemaExpressionParser:
+    (JStreamTableEnv, DataStream[JTuple2[JBool, JTuple5[JLong, JInt, String, JInt, JLong]]]) = {
+
+    val jStreamExecEnv = mock(classOf[JStreamExecEnv])
```
I will remove changes in this file. It not only introduce confusions raised by you, but also can be tested in our `FromUpsertStreamTest`

hequn8128 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/FromUpsertStreamTest.scala:
```diff
@@ -0,0 +1,140 @@
+  }
+
+  @Test
+  def testMaterializeTimeIndicatorAndCalcUpsertToRetractionTranspose() = {
```
`testMaterializeTimeIndicatorAndCalcUpsertToRetractionTranspose` is used to test materialize logic and transpose(calc can be pushed down through UpsertToRetraction).

`testCalcCannotTransposeUpsertToRetraction` is used to test transpose logic(calc can not be pushed down).

However, we should not test transpose logic in this commit? I will remove `testCalcCannotTransposeUpsertToRetraction` in this commit and add it later. What do you think?

hequn8128 replied:
`Calc` can't always be pushed down(More details in `CalcUpsertToRetractionTransposeRule`), so there are two tests for each of the case:
- `testCalcTransposeUpsertToRetraction()`
- `testCalcCannotTransposeUpsertToRetraction()`

As for `testMaterializeTimeIndicatorAndCalcUpsertToRetractionTranspose()`, maybe it's better to rename to `testMaterializeTimeIndicator()`. It is dedicated to test the materialization logic.

What do you think?

pnowojski replied:
Yes, that was my point. If you drop the first condition it will also simplify this test.

hequn8128 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/FromUpsertStreamTest.scala:
```diff
@@ -0,0 +1,140 @@
+  }
+
+  @Test
+  def testCalcCannotTransposeUpsertToRetraction() = {
```
ditto

hequn8128 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/table/validation/TableSinkValidationTest.scala:
```diff
@@ -46,6 +46,20 @@ class TableSinkValidationTest extends TableTestBase {
+    t.writeToSink(new TestAppendSink)
+
+    // must fail because table is not append-only
+    env.execute()
```
Good point! I think we can use `streamTestUtil` to test the sink logic.

hequn8128 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/utils/StreamTestData.scala:
```diff
@@ -95,4 +95,41 @@ object StreamTestData {
+    env.fromCollection(data)
+  }
+
+  def get3TupleUpsertStream(env: StreamExecutionEnvironment):
```
Oh, it should be used in the next commit. I will correct this.

hequn8128 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/StreamTableEnvironmentTest.scala:
```diff
@@ -185,4 +198,20 @@ class StreamTableEnvironmentTest extends TableTestBase {
+    val dsType = new TupleTypeInfo(Types.BOOLEAN, sType)
+      .asInstanceOf[TupleTypeInfo[JTuple2[JBool, JTuple5[JLong, JInt, String, JInt, JLong]]]]
+    val ds = mock(classOf[DataStream[JTuple2[JBool, JTuple5[JLong, JInt, String, JInt, JLong]]]])
+    when(ds.getType).thenReturn(dsType)
```
ditto

hequn8128 commented:
@pnowojski Thanks a lot for your comments. I have updated the PR according to your suggestions and rebased to the master. Looking forward to your new comments. Thank you!

Best, Hequn

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/logical/CalcUpsertToRetractionTransposeRule.scala:
```diff
@@ -0,0 +1,109 @@
+              case ref: RexInputRef =>
+                (ref.getIndex, p.right)
+              case _ =>
+                (-1, p.right)
```
Why do you create a tuples `(int, string)` and later filtering out based on integers instead of for example inserting or not something to an array or a map? Another remark, I think that maybe implementing `RexVisitorImpl` (like `InputRefVisitor`) would be a better, more re-usable approach which would for example solve the problem of duplicating the code for `case r: RexInputRef`.

hequn8128 replied:
Maybe we can use `RexVisitorImpl` in  `FlinkLogicalCalc`?

pnowojski replied:
What do you mean? 

Maybe to rephrase my comment: I think that instead of having this `match case` construct a better idea might to implement it's logic as `RexVisitorImpl`, probably as a private static nested class inside `FlinkLogicalCalc` or `CommonCalc`.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/logical/CalcUpsertToRetractionTransposeRule.scala:
```diff
@@ -0,0 +1,109 @@
+  /**
+    * Get related output field names for input field names for Calc.
+    */
+  private def getInOutNames(calc: FlinkLogicalCalc): Seq[(String, String)] = {
```
Some kind of `map` usually better expresses mappings between two things compared to list of tuples. Is there a reason that I do not see why have you chosen list of tuples?

hequn8128 replied:
ditto

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/logical/CalcUpsertToRetractionTransposeRule.scala:
```diff
@@ -0,0 +1,109 @@
+    // column pruning or push Filter down
+    calc.getRowType.getFieldCount <= upsertToRetraction.getRowType.getFieldCount &&
+    // key fields should not be changed
+      fieldsRemainAfterCalc(upsertToRetraction.keyNames, calc)
```
I will remove the comments.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/logical/CalcUpsertToRetractionTransposeRule.scala:
```diff
@@ -0,0 +1,109 @@
+  /**
+    * Get related output field names for input field names for Calc.
+    */
+  private def getInOutNames(calc: FlinkLogicalCalc): Seq[(String, String)] = {
```
A field may contain multi-output. For example, `select key as key1, key as key2 from T`.
Maybe refactor it into:
class FlinkLogicalCalc:
```
  /**
   * Returns empty if output field is not forwarded from the input.
   */
  Option[String] getInputFromOutputName(String output)
```
What do you think?

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/logical/FlinkLogicalUpsertToRetraction.scala:
```diff
@@ -0,0 +1,74 @@
+    val rowCnt = mq.getRowCount(child)
+    // take rowCnt and fieldCnt into account, so that cost will be smaller when generate
+    // UpsertToRetractionConverter after Calc.
+    planner.getCostFactory.makeCost(rowCnt, rowCnt * child.getRowType.getFieldCount, 0)
```
I think we can make the cost same with source, e.g., `FlinkLogicalNativeTableScan`. What do you think?

pnowojski replied:
As far as I can see `FlinkLogicalNativeTableScan` is the same to what I proposed in the comment above, right? The only difference is `estimateRowSize(child.getRowType)` vs `estimateRowSize(getRowType)`, which probably both are equivalent in this case.

hequn8128 replied:
Yes, I think so. I updated the code according to your suggestion.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/logical/FlinkLogicalUpsertToRetraction.scala:
```diff
@@ -0,0 +1,74 @@
+import org.apache.flink.table.plan.logical.rel.LogicalUpsertToRetraction
+import org.apache.flink.table.plan.nodes.FlinkConventions
+
+class FlinkLogicalUpsertToRetraction(
```
Only calc don't remove key fields can be pushed down. We make sure this in `CalcUpsertToRetractionTransposeRule`. After optimization, keyNames will be renames according to the calc.

pnowojski replied:
`FlinkLogicalUpsertToRetraction` has hidden `RexNode` reference - in form of `keyNames`.

I think currently calcite is only using `accept()` or `getChildExps()` for decorrelating the plan. So currently there is no problem, because `FlinkLogicalUpsertToRetraction.keyNames` should never have anything to do with corellated subqueries. But my point is that if someone would use those methods for something else, like renaming variables, looking which fields are referenced, etc..., it would brake very silently and the failure would be very hard to debug.

Look at the all of the implementations of `org.apache.calcite.rel.RelNode#accept(org.apache.calcite.rex.RexShuttle)`. Basically every node that has some some input references (or other expressions), is implementing this method.

Unfortunately you might be right, that throwing an exception will not work here :/

hequn8128 replied:
Thanks for your reply. I guess we are on the same page now?

pnowojski replied:
I'm syncing with Calcite dev mailing list, trying to understand those methods. Until we get a clear response from them lets leave this as it is now.

hequn8128 replied:
Thank you very much for looking into this. I will also keep an eye on the Calcite dev mailing list.

pnowojski replied:
They have responded to me and basically there is no contract... as long as the tests are passing we are fine. I don't like it, but if you want we can leave those methods not implemented.

hequn8128 commented:
@pnowojski Thanks again for the suggestions. I have updated the pr and rebased to the master. Thank you.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/CommonCalc.scala:
```diff
@@ -19,18 +19,49 @@
+  /**
+    * Returns empty if output field is not forwarded from the input for the calc.
+    */
+  private[flink] def getInputFromOutputName(calc: Calc, outputFieldName: String): Option[String] = {
```
Could you add a simple unit test for this?

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/logical/CalcUpsertToRetractionTransposeRule.scala:
```diff
@@ -0,0 +1,82 @@
+
+  private def fieldsRemainAfterCalc(fields: Seq[String], calc: FlinkLogicalCalc): Boolean = {
+    calc.getRowType.getFieldNames
+      .flatMap(calc.getInputFromOutputName(calc, _))
```
[newbie scala question]
How does it work? `flatMap` here is unwrapping `Option` returned from `getInputFromOutputName`? If `Option` is empty, the result list will have null, will it be empty or will it throw an exception? (two first options are acceptable ;) )
[/newbie scala question]

hequn8128 replied:
The `flatMap` unwraps `Option` returned from the `getInputFromOutputName`. If `Option` is empty, the result will be an empty list. :-)

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/StreamScan.scala:
```diff
@@ -19,22 +19,138 @@
+    }
+  }
+
   protected def convertToInternalRow(
```
OK, thanks for your explanation. Make sense. I will rename this method to `convertAppendToInternalRow` keeping it in sync with `convertUpsertToInternalRow`.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/StreamScan.scala:
```diff
@@ -19,22 +19,138 @@
 
 trait StreamScan extends CommonScan[CRow] with DataStreamRel {
 
+  protected def convertUpsertToInternalRow(
```
1)The type of an upsert stream should be scala or java tuple, while we don't have this constraint on append stream.
2&3)Good suggestions. I will try to refactor this making less duplicating logic and more smaller methods.
4)The else branch for java tuple2 hasn't been converted by the tests. I will add one. Thank you.

pnowojski replied:
Ok, I talked also a little bit more with @twalthr and now I understand why do we need this `switch/case`.

Regarding the code duplication, I'm not sure but maybe it could be deduplicated by modifying already existing `convertToInternalRow` and adding two more parameters:
1. `Row getRow(T inputRecord)`
2. `boolean getChange(T inputRecord)`

and using those two parameters:
```
new RichMapFunction[T, CRow] {
        ...

        override def map(inputRecord: T): CRow = {
          outCRow.row = getRow(inputRecord)
          outCRow.change = getChange(inputRecord)
          outCRow
        }
}
```

and this modified `convertToInternalRow` method could be called three times:
1. for the current use case of append only data stream with `getRow(record) : record`, `getChange(record) : true`
2. for upsert scala: with `getRow(record) : record._2`, `getChange(record) : record._1`
3. for upsert java: with `getRow(record) : record.f1`, `getChange(record) : record.f0`

or something along those lines.

hequn8128 replied:
Thanks a lot for the ideas. I will try this and give some feedback later.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/StreamTableEnvironment.scala:
```diff
@@ -426,7 +426,7 @@ abstract class StreamTableEnvironment(
 
       case Some(func) =>
-        new CRowMapRunner[OUT](func.name, func.code, func.returnType)
+        new CRowToExternalTypeMapRunner[OUT](func.name, func.code, func.returnType)
```
You can keep this rename and class moving as a separate `[hotfix]` commit either at the beginning or at the end of this PR or as separate PR :) We do not have to drop those changes.

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/UpsertToRetractionProcessFunction.scala:
```diff
@@ -0,0 +1,104 @@
+  * @param rowTypeInfo the output row type info.
+  * @param queryConfig the configuration for the query.
+  */
+class UpsertToRetractionProcessFunction(
```
Ok, but then we can provide `LAST_NULLABLE_VALUE()` function and still re-use aggregation operator code. Even if we keep `LAST_NULLABLE_VALUE()` as an internal function that we do not expose anywhere. I would vote for exposing it (maybe in a follow up PR).

My main concern is again code duplication with aggregation operator. 

Now that I'm thinking about it, did we even need to implement our own UpsertToRetraction rules and RelNodes? Couldn't we re-use `Aggregate` `RelNode` with `LAST_NULLABLE_VALUE()` aggregation functions?

hequn8128 replied:
I think your concern makes sense. However, I'm not sure if we can use the `LAST_NULLABLE_VALUE` function. According to the description of handling null, it is said: `aggregate functions ignore null values`. This means we can even skip processing the data. Some optimization can make use of it to gain better performance. Anything like this would be broken silently later. 
Again, I agree with you that it is good to reuse the aggregation code, so if you do think it is valuable to do. I can change the current implementation.

pnowojski replied:
> it is said: `aggregate functions ignore null values`

Can you point me where did you get this? Is this part of SQL standard? `COUNT` function definitely doesn't ignore null values, so I would assume we should be able to define custom aggregate functions that work in a similar fashion. [This seems to support this](
https://modern-sql.com/concept/null#aggregates):
> Most aggregate functions that return structured data (array_agg, json_objectagg, xmlagg) do not remove null values.
> ...
> For array_agg, it is even explicitly mentioned in SQL:2016-2: §10.9, General Rule 12gii, note 510.

hequn8128 replied:
I haven't found related descriptions in SQL standard. I found these descriptions in some popular sql systems, such as[[1]](https://docs.microsoft.com/en-us/sql/t-sql/functions/aggregate-functions-transact-sql?view=sql-server-2017)[[2]](https://docs.oracle.com/database/121/SQLRF/functions003.htm#SQLRF20035).

>Except for COUNT, aggregate functions ignore null values. 

I'm fine if we decide to reuse aggregate operator and I prefer reuse all aggregate operator including planing instead only runtime part to make it more clear. What do you think?

pnowojski replied:
The only reference in SQL standard that I have found was this `SQL:2016-2: §10.9, General Rule 12gii, note 510.`, which explicitly states that `array_agg` can not ignore nulls. Also I haven't found in the standard  references to something stating the opposite, so I would be inclined to re-use aggregation node in planning as well if you are fine with that.

Next question would be wether to simply create an aggregation node or to extend either `Aggregate` or `FlinkLogicalAggregate` classes (to re-use all of their logic), but to keep the distinct name of the node `UpsertToRetraction`?

hequn8128 replied:
OK. Maybe we can simply create a `FlinkLogicalAggregate` in `UpsertToRetractionConverter`, thus we can reuse the rules of Aggregate such as `FilterAggregateTransposeRule`?

pnowojski replied:
Just FYI. `FilterAggregateTransposeRule` should work with any class that extends `Aggregate` node, so if our `FlinkLogicalUpsertToRetraction` class would extend from `FlinkLogicalAggregate`, transpose rule should be fine.

Creating `FlinkLogicalAggregate` in `UpsertToRetractionConverter` is one option, but with a slight drawback of less easier to understand query plan. If logical/physical query plan has an `UpsertToRetraction` node it's obvious what it does. If there is an `Aggregate` node with some `LAST_NULLABLE_VALUE` function, it might be confusing "where did it come from?". However this point may be addressed by overriding `RelNode#getRelTypeName` in `FlinkLogicalAggregate`, so that the aggregation created by `UpsertToRetractionConverter` would return `"FlinkLogicalUpsertToRetraction"` name.

hequn8128 replied:
Sorry for my typo. I meant to create `LogicalAggregate` in `UpsertToRetractionConverter`.  Some rules such as `AggregateProjectPullUpConstantsRule` only matches `LogicalAggregate`. Thus we can reuse more Aggregate rules for `UpsertToRetraction` node. 

To make our plan more understandable, we may can't override the `getRelTypeName` method. It is final in `AbstractRelNode`. However, we may override `explainTerms` and `toString` method to make the RelNode more understandable.
What do you think?

pnowojski replied:
Ops, I missed this `final` :/ Will overriding `explainTerms` or `toString` help with anything? I think calcite's explain (`org.apache.calcite.rel.externalize.RelWriterImpl#explain_`) hard codes using `getRelTypeName()` and we are using this for printing the plans, or am I missing something? 

Ok, please do as you decide is best. It would be nice to keep explained plan easy to understand, but if we just use plain aggregation for the sake of code simplicity, it's not the end of the world.

hequn8128 replied:
OK, I will try to reimplement the PR and reuse the Aggregate node. I will let you know once it is ok. Thanks a lot for your valuable suggestions. :-)

pnowojski commented on flink-examples/flink-examples-table/src/main/scala/org/apache/flink/table/examples/scala/StreamTableExample.scala:
```diff
@@ -44,12 +44,12 @@ object StreamTableExample {
     val orderA = env.fromCollection(Seq(
```
nit: 2nd commit
> [FLINK-8577][table] Rename methods name to toTableFromAppendStream and toTableFromUpsertStream in DataStreamConversions

has no code changes related to `toTableFromUpsertStream`. Probably commit should be renamed.

hequn8128 replied:
Will rename it to `Rename methods name to toTableFromAppendStream in DataStreamConversions`

pnowojski commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/StreamTableEnvironment.scala:
```diff
@@ -555,18 +555,102 @@ abstract class StreamTableEnvironment(
+    val (fieldNames, fieldIndexes) = getFieldInfo[T](streamType, fields)
+
+    // validate and extract time attributes
+    val (rowtime, proctime) = validateAndExtractTimeAttributes(streamType, fields)
```
I think rowtime columns shouldn't be allowed for upsert streams at the moment.

hequn8128 replied:
OK, we can disable it in this PR.

hequn8128 commented:
@pnowojski I'm back from the Vocation. Sorry for the long silence. Thank you very much for your valuable comments. I have replied to your comments correspondingly. Would be great if you can take a look.
May I ask a question? Do we have any plan on merging this PR? Is it possible to make this feature available in 1.8  :-)

Best, Hequn

pnowojski commented:
Hi @hequn8128, sorry for not responding for 2 weeks, but this time I was on the holidays :(

hequn8128 commented:
@pnowojski I have reimplemented the PR and reused the Aggregate node. However, when I try to add the transpose logic I find we need a more generic rule to transpose Calc and Aggregate. Maybe we can add a rule in Calcite, so could I remove the transpose logic from the current PR and implement it later? This can also help us to make the current PR smaller thus easier to be reviewed and merged. 
What do you think?

pnowojski commented:
Yes, this sounds good. I see that this rule can be more difficult to implement for a generic `Aggregate` node.

tobegit3hub commented:
Any update for this PR? We are really looking for these APIs to convert append stream or upsert stream back to Table objects.

rex-remind101 commented:
Any update here? We could very much use this. Thanks!

wuchong commented:
Hi @rex-remind101 , this feature has been consumed by [FLIP-149](https://cwiki.apache.org/confluence/display/FLINK/FLIP-149%3A+Introduce+the+upsert-kafka+Connector) and [FLIP-136](https://cwiki.apache.org/confluence/display/FLINK/FLIP-136%3A++Improve+interoperability+between+DataStream+and+Table+API). 

FLIP-149 has been implemented in 1.12 and FLIP-136 will be supported in the next 1.13 version.

I will close this issue and PR then.
